### DISCUSSION TITLE
Lazy loading collections (one-to-many, many-to-many)

### DISCUSSION THREAD

**Discussion by gusgorman2** [mikro-orm/mikro-orm#4934] [General]
Hello,

Is there a way to configure a one-to-many or many-to-many relationship to be lazy loaded automatically by default?

For example, if we have `order.orderItems`,  `orderItems` would be loaded when we access that property without having to call `order.orderItems.loadItems()` ?

I don't want to have to pollute my domain logic with any knowledge about whether something has already been loaded yet or not.

**Comment by B4nan** [comment]
Nope, touching the database is async operation, so it will always require some async method, no way around that in JS.

**Reply by gusgorman2** [reply]
Ok - i realise I'm into feature request territory now, but could there be a way of specifying the relationship in the first place that means that the property is async by default ?

**Reply by B4nan** [reply]
wdym? can you provide some example?

**Reply by gusgorman2** [reply]
something like:

```
  @OneToMany(() => OrderItem, (o) => o.orderItem)
  orderItems:  Promise<Collection<OrderItem>>
```

I only ask because this sort of lazy loading is fundamental in ORMS in Java or .NET, and means you can write domain logic without polluting it with any knowledge as to what is or isn't loaded.

I realise JS is s very different beast, but just wondered if it is possible?

**Reply by B4nan** [reply]
Possible, but I doubt that would be a good idea, I am not interested in going that way. I don't see how this would be better than using the `loadItems` method. In a way you are already "polluting" your domain with the `Collection` class, you won't get around that even with your approach.

I'd also argue that relying on lazy loading is the mother of all evil (= N+1 selects). You should prefer preloading things, the ORM has a smart typing (`Loaded` type) that will respect your loading hints and expose type-safe `$` properties on the loaded relations, that way you can write your code so it requires relations to be loaded on type-level and it will fail to compile if you try to pass something without the loaded relation hint.

https://mikro-orm.io/docs/next/guide/type-safety

**Reply by gusgorman2** [reply]
I guess we just come at this from a different angle. I have made extensive use of lazy loading on many projects over the last >15 years, and on the rare occasions that I've encountered N+1 select problems have dealt with them by tweaking that particular mapping, or writing a query for that particular case.

On the other hand I find the idea of having to prescriptively preload data brittle and inflexible.

Of course i appreciate its your project - I'll just have to work with it ! :)
